### DISCUSSION TITLE
arm64 docker images

### DISCUSSION THREAD

**Discussion by bjo81** [pretix/pretix#2681] [Ideas]
Hi,
would it be possible to provide arm64 docker images?

Regards

**Comment by raphaelm** [comment]
I never looked into how much extra work that would be, I'm mostly worried about publishing a version of the image we don't use ourselves and hence do not regularly test.

It should be possible, though, at least ~3-4 years ago pretix ran fine on a raspberry pi, so  most dependencies should be arm-compatible, hoping that did not change.

One problem might be our base image is based on the official python image which is based on Debian which in turn does not officially support arm, does it?

**Reply by bjo81** [reply]
I tried to build on a Oracle Cloud Free Tier arm64 box as it has a lot of memory, `python:3.9-bullseye` seems to exist for arm64, but rust is missing:

```
Building wheels for collected packages: css-inline, django-formset-js-improved, libsass, paypal-checkout-serversdk, paypalrestsdk, python-u2flib-server, static3, vobject, dj-static, mysqlclient, pycountry, pylibmc, slimit, future, pyrsistent, django-jquery-js
  Building wheel for css-inline (pyproject.toml): started
  Building wheel for css-inline (pyproject.toml): finished with status 'error'
  error: subprocess-exited-with-error
  
  × Building wheel for css-inline (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [15 lines of output]
      running bdist_wheel
      running build
      running build_ext
      running build_rust
      error: can't find Rust compiler
      
      If you are using an outdated pip version, it is possible a prebuilt wheel is available for this package but pip is not able to install from it. Installing from the wheel would avoid the need for a Rust compiler.
      
      To update pip, run:
      
          pip install --upgrade pip
      
      and then retry package installation.
      
      If you did intend to build this package from source, try installing a Rust compiler from your system package manager and ensure it is on the PATH during installation. Alternatively, rustup (available at https://rustup.rs) is the recommended way to download and update the Rust compiler toolchain.
      [end of output]
```

**Reply by raphaelm** [reply]
:( Then it's probably not going to be simple

**Reply by bjo81** [reply]
`cargo` and `rustc` seems to be available for arm64 in the bullseye repo, but cargo seems to be too old.

**Reply by bjo81** [reply]
Tried to build it with rust from rustup, fails now with:
```
 error: cannot find macro `pyo3_built` in this scope                                                                                                                                     
         --> src/lib.rs:196:29                                                                                                                                                                
          |                                                                                                                                                                                   
      196 |     module.add("__build__", pyo3_built::pyo3_built!(py, build))?;                                                                                                                 
          |                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^                                                                                                                    
          |                                                                                                                                                                                   
          = note: `pyo3_built` is in scope, but it is a crate, not a macro                                                                                                                    
          = note: this error originates in the macro `pyo3_built::pyo3_built` (in Nightly builds, run with -Z macro-backtrace for more info)                                                  
                                                                                                                                                                                              
      error: could not compile `css-inline-python` due to previous error     
``` 
:thinking:

**Reply by Stranger6667** [reply]
The issue above is fixed in newer `css-inline` versions + from `0.8.5` there are many more binary wheels, including `aarch64` on musl / glibc

**Reply by yvovandoorn** [reply]
> The issue above is fixed in newer `css-inline` versions + from `0.8.5` there are many more binary wheels, including `aarch64` on musl / glibc

@Stranger6667 yup. PR #2818 fixed it. Thanks for following up 🙂

**Comment by yvovandoorn** [comment]
Bumping `css-inline` from `css-inline==0.7.*` to `css-inline==0.8.*` installs Pretix w/o an issue (tested by git clone). This is because with `0.7.8` there is no wheel available on pypi for arm64 as the maintainer only started generating arm64 wheels with `0.8.0.` So what happens is it has to build from source:

```
Collecting css-inline==0.7.*
  Using cached css_inline-0.7.8.tar.gz (21 kB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... canceled
```

This then requires rust and eventually has a compilation error. 

With `css-inline 0.8.*` specified, this happens:
```
Collecting css-inline==0.8.*
  Downloading css_inline-0.8.3-cp310-cp310-manylinux_2_17_aarch64.manylinux2014_aarch64.whl (2.1 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.1/2.1 MB 56.2 MB/s eta 0:00:00
```

**Comment by yvovandoorn** [comment]
For those that are going to be adventurous with an arm64 installation of Pretix. I moved over the self hosted installation I support on major cloud provider yesterday from x86 to arm64. During the migration, when I tried to run `python -m pretix migrate` it gave the following error:

```
SCRAM authentication requires libpq version 10 or above
```

This is due to the fact that on arm64 & debian based distros the python package `psycopg2-binary` is somehow linking against libpq v9, which doesn't support SCRAM authentication. 

Solution:

Run `pip install psycopg2` as the pretix user. This will compile `psycopg2` against the system version of libpq (you might need to run `apt install -y libpq-dev` which is going to be version 13 or higher.

**Reply by marcoacierno** [reply]
Do you still run Pretix on ARM? Did you have other issues or similar in the last ~2 years?

Thanks for sharing your experience!

**Comment by yvovandoorn** [comment]
Still happily using it on arm64 with the above workaround in place.
